src/components/timeline/Timeline.svelte

```diff
@@ -21,35 +43,125 @@
+    tickCount = clamp(ticks, 2, 15);
+  }
+
   $: setRowsMaxHeight(timelineDiv, xAxisDiv, timelineHistogramDiv);
   $: xDomainMax = [new Date($maxTimeRange.start), new Date($maxTimeRange.end)];
   $: xDomainView = [new Date($viewTimeRange.start), new Date($viewTimeRange.end)];
   $: xScaleMax = getXScale(xDomainMax, drawWidth);
   $: xScaleView = getXScale(xDomainView, drawWidth);
-  $: xTicksView = xScaleView.ticks(tickCount).map((date: Date) => {
-    const doyTimestamp = getDoyTime(date, false);
-    const [yearDay, time] = doyTimestamp.split('T');
-    return { date, time, yearDay };
-  });
+  $: xScaleViewDuration = $viewTimeRange.end - $viewTimeRange.start;
+
+  // Use a custom D3 time day to force equidistant time intervals
+  // for days as opposed to D3's non-uniform intervals that can end early
+  // on months or years
+  // See https://github.com/d3/d3-scale/issues/245
+  // And https://observablehq.com/d/906f777c9f2f0701
+  const customD3TimeDay = timeInterval(
+    date => date.setHours(0, 0, 0, 0),
+    (date, step) => date.setDate(date.getDate() + step),
+    (start, end) => (end.getTime() - start.getTime()) / durationDay,
+    date => Math.floor(date.getTime() / durationDay),
+  );
+
+  // TODO may need custom hour, week, month?
+
+  // From https://github.com/d3/d3-time/blob/main/src/ticks.js
+  const customD3TickIntervals: [CountableTimeInterval, number, number][] = [
+    [timeSecond, 1, durationSecond],
+    [timeSecond, 5, 5 * durationSecond],
+    [timeSecond, 15, 15 * durationSecond],
+    [timeSecond, 30, 30 * durationSecond],
+    [timeMinute, 1, durationMinute],
+    [timeMinute, 5, 5 * durationMinute],
+    [timeMinute, 15, 15 * durationMinute],
+    [timeMinute, 30, 30 * durationMinute],
+    [timeHour, 1, durationHour],
+    [timeHour, 3, 3 * durationHour],
+    [timeHour, 6, 6 * durationHour],
+    [timeHour, 12, 12 * durationHour],
+    [customD3TimeDay, 1, durationDay],
+    [customD3TimeDay, 2, 2 * durationDay],
+    [timeWeek, 1, durationWeek],
+    [timeMonth, 1, durationMonth],
+    [timeMonth, 3, 3 * durationMonth],
+    [timeYear, 1, durationYear],
+  ];
+
+  // From https://github.com/d3/d3-time/blob/main/src/ticks.js
+  function customD3TickInterval(start: number, stop: number, count: number) {
+    const target: number = Math.abs(stop - start) / count;
+    const i = bisector(([, , step]) => step).right(customD3TickIntervals, target);
+    if (i === customD3TickIntervals.length) {
+      return timeYear.every(tickStep(start / durationYear, stop / durationYear, count));
+    }
+    if (i === 0) {
+      return timeMillisecond.every(Math.max(tickStep(start, stop, count), 1));
+    }
+    const [t, step] =
+      customD3TickIntervals[
+        target / customD3TickIntervals[i - 1][2] < customD3TickIntervals[i][2] / target ? i - 1 : i
+      ];
+    return t.every(step);
+  }
+
+  // From https://github.com/d3/d3-time/blob/main/src/ticks.js
+  function customD3Ticks(start, stop, count) {
```

Comment:
Are we able to move this to `utilities/timeline.ts`?
Also can we strongly type these parameters?

Reply:
Yep

Reply:
Well, I'll try, not always super clear how to type these internal D3 bits as their types are stored separately.

Reply:
Might be some funny business with some of the D3 types as well, seeing some issues with some of their non-typed examples